### PR TITLE
samplescan operators for Myria

### PR DESCRIPTION
**Pull request by radion**
@bmyerz I'll be adding the generic sample(expression) later as well. This code generates proper plans for sampling from a base relation.
I'll probably change the scanner/parser later to make the MyriaL function more user-friendly, but the logic in this PR will stay the same.
Let me know any feedback here. Thanks!

### CODE REVIEW

File: raco/myrial/scanner.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 builtins = ['EMPTY', 'WORKER_ID', 'SCAN', 'COUNTALL', 'COUNT', 'STORE',
             'DIFF', 'CROSS', 'JOIN', 'UNIONALL', 'INTERSECT', 'DISTINCT',
-            'LIMIT', 'SINK']
+            'LIMIT', 'SINK', 'SAMPLEWRSCAN', 'SAMPLEWORSCAN']
```
Comment (bmyerz) [inline raco/myrial/scanner.py]: These need less opaque names. Why not make just SAMPLESCAN with an extra argument?

File: raco/myrial/interpreter.py [hunk 2]
```diff
@@ -97,6 +97,19 @@ def scan(self, rel_key):
...
+            scheme = self.catalog.get_scheme(rel_key)
+        except KeyError:
+            if not self.use_dummy_schema:
+                raise NoSuchRelationException(rel_key)
+
+            # Create a dummy schema suitable for emitting plans
+            scheme = raco.scheme.DummyScheme()
+
+        return raco.algebra.SampleScan(rel_key, scheme, samp_size, is_WR)
+
```
Comment (bmyerz) [inline raco/myrial/interpreter.py]: seems like it should share code with scan above, e.g. `_get_scan_scheme`

File: raco/language/myrialang.py [hunk 3]
```diff
@@ -796,12 +803,102 @@ def compileme(self):
...
+
+    def compileme(self, leftid, rightid):
+        return {
+            "opType": "Sample",
+            "argChild1": leftid,
+            "argChild2": rightid,
+        }
+
+
+class AddSampleOperators(rules.Rule):
```
Comment (bmyerz) [inline raco/language/myrialang.py]: More descriptive name, like SampleToDistributedSample

File: raco/language/myrialang.py [hunk 4]
```diff
@@ -796,12 +803,102 @@ def compileme(self):
         }
 
 
+class MyriaSamplingDistribution(algebra.UnaryOperator, MyriaOperator):
```
Comment (bmyerz) [inline raco/language/myrialang.py]: More descriptive name is `MyriaCalculateSampling(Parameters|Distribution)`?

File: raco/language/myrialang.py [hunk 5]
```diff
@@ -632,7 +639,7 @@ def num_tuples(self):
     def compileme(self, inputid):
         if len(self.hash_columns) == 1:
             pf = {
-                "type": "SingleFieldHash",
+                "type": "RawValue" if self.is_raw_val else "SingleFieldHash",
```
Comment (bmyerz) [inline raco/language/myrialang.py]: What is `RawValue` indicating to MyriaX?

File: raco/algebra.py [hunk 6]
```diff
@@ -1363,6 +1365,38 @@ def scheme(self):
...
+class SampleScan(ZeroaryOperator):
+
+    """Logical Sample Operator"""
+
+    def __init__(self, relation_key, _scheme, sample_size, with_replacement):
+        self.relation_key = relation_key
+        self._scheme = _scheme
+        self.sample_size = sample_size
+        self.with_replacement = with_replacement
+        self.sample_type = "WR" if self.with_replacement else "WoR"
```
Comment (bmyerz) [inline raco/algebra.py]: this field appears to be redundant with `self.with_replacement`. Can you just have a private method to return the appropriate one?

File: raco/algebra.py [hunk 7]
```diff
@@ -1006,9 +1006,10 @@ class Shuffle(UnaryOperator):
 
     """Send the input to the specified servers"""
 
-    def __init__(self, child=None, columnlist=None):
+    def __init__(self, child=None, columnlist=None, raw_value_shuffle=False):
         UnaryOperator.__init__(self, child)
         self.columnlist = columnlist
+        self.raw_value_shuffle = raw_value_shuffle
```
Comment (bmyerz) [inline raco/algebra.py]: see comment below

File: raco/language/myrialang.py [hunk 8]
```diff
@@ -796,12 +803,102 @@ def compileme(self):
...
+class MyriaSample(algebra.BinaryOperator, MyriaOperator):
+    """A Myria Sample operator"""
+    def __init__(self, left, right, sample_size, with_replacement):
+        algebra.BinaryOperator.__init__(self, left, right)
+        # with_replacement and sample_size are just used for displaying.
+        self.with_replacement = with_replacement
+        self.sample_size = sample_size
+
+    def shortStr(self):
+        samp_type = "WR" if self.with_replacement else "WoR"
```
Comment (bmyerz) [inline raco/language/myrialang.py]: Make this consistent with the other `sample_type` code. Probably put the `"WR"` and `"WoR"` in exactly one place in the code

**Comment by bmyerz** [comment]
- Add tests
- Add one example to `examples/` and `examples/catalog.py`

File: raco/language/myrialang.py (hunk 5 above)
Comment (radion) [inline raco/language/myrialang.py]: I added a new type of Shuffling to MyriaX - a shuffle that sends to the worker id matching a raw attribute value (i.e. the non-hashed value). RACO needs to be aware of this.

File: raco/myrial/scanner.py (hunk 1 above)
Comment (radion) [inline raco/myrial/scanner.py]: I considered making something like samplescan(relation, 500, WR) and samplescan(relation, 500, WoR). Would it be possible to detect the strings 'WR' or 'WoR' in the parser?

Comment (bmyerz) [inline raco/myrial/scanner.py]: yes, there is a `STRING_LITERAL` token

File: raco/language/myrialang.py (hunk 5 above)
Comment (bmyerz) [inline raco/language/myrialang.py]: Better name might be `IdentityHash` or something like that. I didn't know what RawValue meant by reading it

**Comment by radion** [comment]
@bmyerz the latest few commits should cover the feedback you gave me.

File: raco/language/myrialang.py [hunk 9]
```diff
@@ -611,14 +612,29 @@ def compileme(self, inputid):
...
         algebra.UnaryOperator.__init__(self, input)
         self.hash_columns = hash_columns
+        # If no specified shuffle type, it's a single/multi field hash.
+        if shuffle_type is None:
+            if len(self.hash_columns) == 1:
+                self.shuffle_type = ShuffleType.SingleFieldHash
+            else:
+                self.shuffle_type = ShuffleType.MultiFieldHash
+        else:
+            self.shuffle_type = shuffle_type
```
Comment (bmyerz) [inline raco/language/myrialang.py]: Add an assertion that `len(hash_columns) == 1` if the type is IdentityHash

Comment (radion) [inline raco/language/myrialang.py]: Done.

File: raco/language/myrialang.py [hunk 10]
```diff
@@ -611,14 +612,30 @@ def compileme(self, inputid):
         }
 
 
+class ShuffleType(object):
+    """Enum of supported shuffling types."""
+    SingleFieldHash, MultiFieldHash, IdentityHash = range(3)
+
```
Comment (bmyerz) [inline raco/language/myrialang.py]: put this class inside of algebra.Shuffle

File: raco/myrial/exceptions.py [hunk 11]
```diff
@@ -162,3 +162,11 @@ def __init__(self, relname):
 
     def __str__(self):
         return "No such relation: %s" % self.relname
+
+
+class InvalidSamplingTypeException(MyrialCompileException):
```
Comment (bmyerz) [inline raco/myrial/exceptions.py]: Not sure this merits a new expression, why not just `raise` a `MyrialCompileException` with that error message

File: raco/myrial/interpreter.py [hunk 12]
```diff
@@ -82,21 +82,31 @@ def __lookup_symbol(self, _id):
...
+        assert isinstance(rel_key, relation_key.RelationKey)
+        scheme = self._get_scan_scheme(rel_key)
         return raco.algebra.Scan(rel_key, scheme,
                                  self.catalog.num_tuples(rel_key))
 
+    def samplescan(self, rel_key, samp_size, is_pct, samp_type):
+        """Sample a base relation."""
+        assert isinstance(rel_key, relation_key.RelationKey)
+        if samp_type not in ('WR', 'WoR'):
+            raise InvalidSamplingTypeException(samp_type)
```
Comment (bmyerz) [inline raco/myrial/interpreter.py]: see above

**Comment by bmyerz** [comment]
@radion looks good. Couple more changes and I'll merge it

**Comment by senderista** [comment]
Looks like the `sample()` operator never made it in. I thought that the expected plans generated by `sample()` were equivalent to https://github.com/uwescience/myria/blob/master/jsonQueries/radion_queries/SampleWR.json and https://github.com/uwescience/myria/blob/master/jsonQueries/radion_queries/SampleWoR.json, but those plans just seem to be `samplescan()` with presampling.

**Pull request merged**